bin/utils.js

```diff
@@ -0,0 +1,104 @@
+  // https://stackoverflow.com/a/14032965/1327178
+  [`SIGINT`, `SIGTERM`, `SIGUSR1`, `SIGUSR2`].forEach((event) => {
+    process.on(event, () => {
+      console.log(`\n${event} received`);
+      exitCallback(0);
+    });
+  });
+  process
+    /* NodeJS process will exit when:
+     * - the event loop's queue is empty.
+     * - no background/asynchronous tasks remain that are capable of adding to the queue. */
+    .on('exit', () => {
+      console.log('NodeJS has nothing more to execute ...');
+      exitCallback();
+    })
+    .on('uncaughtException', (error, origin) => {
+      console.log(`uncaughtException received`, { error: serializeError(error), origin });
+      exitCallback();
+    })
+    .on('unhandledRejection', (reason) => {
+      console.log(`unhandledRejection received`, { reason });
+      exitCallback();
+    });
+};
+
+// TODO: load from config if flexibility is required
+const JOB_EXPIRATION_DURATION = dayjs.duration(1, 'hour');
+const JOB_EXPIRATION_DURATION_STR = JOB_EXPIRATION_DURATION.humanize();
+
+const CLEANUP_INTERVAL_MS = _.toNumber(ms('1 hour'));
+const CLEANUP_INTERVAL_STR = dayjs.duration(CLEANUP_INTERVAL_MS, 'ms').humanize(true);
+
+const cleanupStaleJobs = (agenda) => {
+  const cleanup = () => {
+    void (async () => {
+      console.log(`Cleaning-up completed one-shot jobs older than ${JOB_EXPIRATION_DURATION_STR} ...`);
+
+      try {
+        let cleanupCount = 0;
+        let totalJobCount = 0;
+
+        const collection = agenda._collection || agenda._collection.collection;
+        if (!collection) {
+          console.log(`Cannot perform cleanup: agenda.collection not found!`);
+          return;
+        }
+
+        const now = dayjs();
+        const expirationTime = now.subtract(JOB_EXPIRATION_DURATION);
+
+        await collection.find({ type: 'normal' }).forEach((job) => {
+          totalJobCount++;
+          const jobId = job._id;
+
+          if (!job.lastFinishedAt) {
+            // skip job since it hasn't finished yet
+            return;
+          }
+          const lastFinishedAt = dayjs(job.lastFinishedAt);
+          if (!lastFinishedAt.isValid()) {
+            console.log(`[${job._id}] Skipping job due to invalid lastFinishedAt date`);
+            return;
+          }
+
+          if (lastFinishedAt.isBefore(expirationTime)) {
+            console.log(`[${jobId}] Deleting ${job.name} (job finished ${JOB_EXPIRATION_DURATION_STR} ago)`);
+            agenda.cancel({
+              _id: ObjectId(jobId),
+            });
+            cleanupCount++;
+          }
+        })
+
+        console.log(`Cleaned-up ${cleanupCount}/${totalJobCount} one-shot job(s)! Next cleanup runs ${CLEANUP_INTERVAL_STR}.`);
+      } catch (e) {
+        console.log(`Stale job cleanup procedure failed with exception: `, { error: serializeError(e) })
+      }
+    })();
+
+    setTimeout(cleanup, CLEANUP_INTERVAL_MS);
+  };
+
+  agenda.on("ready", async () => {
```

Comment:
Not critical: would `once` be a better fit here? But I'm pretty sure we have guarantees of `ready` only firing once.

Reply:
Eh, can confirm this only gets called once on Agenda object instantiation so... no biggy, can ignore